src/index.js

```diff
@@ -19,177 +19,264 @@ const debug = (function () {
             await this.signout()
           }
         } else if (e.newValue) {
-          await this.connectToMasq()
-          this.eventTarget.dispatchEvent(new Event('logged_in'))
+          // if we are logging in in the current tab
+          if (this.state !== 'authorized' &&
+              this.state !== 'accessMaterialReceived' &&
+              this.state !== 'logged') {
+            await this._loadSessionInfo()
+            debug('[masq._listenForLoginOrSignout] detected login from another window/tab')
+            debug('[masq._listenForLoginOrSignout] dispatch logged_in')
+            this.eventTarget.dispatchEvent(new Event('logged_in'))
+          }
         }
       }
     })
   }
 
   _listenForFocus () {
-    document.addEventListener('focus', () => {
-      this.connectToMasq()
+    document.addEventListener('focus', async () => {
+      // try to reopen db
+      if (!this.isLoggedIn()) {
+        return
+      }
+      debug('[masq._listenForFocus] detected refocus on window')
+      await this._openDb()
+      try {
+        this.startReplication()
+      } catch (e) {
+        console.log('replication after refocus failed')
+      }
     })
   }
 
-  _createSwarm (hub) {
-    return swarm(hub, {
-      config: this.config.swarmConfig
-    })
-  }
+  async _loadSessionInfo () {
+    debug('[masq._loadSessionInfo]')
+    const auxLoginFromUserInfo = async (currentUserInfo) => {
+      const { userAppDbId, userAppDEK, userAppNonce, username, profileImage } = JSON.parse(currentUserInfo)
+      this.userAppDbId = userAppDbId
+      this.userAppDEK = userAppDEK
+      this.importedUserAppDEK = await common.crypto.importKey(Buffer.from(userAppDEK, 'hex'))
+      this.userAppNonce = userAppNonce
+      this.username = username
+      this.profileImage = profileImage
+      await this._openDb()
+      await this.setState('logged')
+    }
 
-  _reset () {
-    this.userAppRepSW = null
-    this.userAppRepHub = null
+    // If user info is stored in session storage, do not use localStorage
+    const currentUserInfo = window.sessionStorage.getItem(CURRENT_USER_INFO_STR)
+    if (currentUserInfo) {
+      await auxLoginFromUserInfo(currentUserInfo)
+      return
+    }
 
-    this._logIntoMasqErr = null
-    this._logIntoMasqDone = false
+    // if userId is is not in session storage, look for it in local storage
+    // and save in session storage
+    const localStorageCurrentUserInfo = window.localStorage.getItem(CURRENT_USER_INFO_STR)
+    if (localStorageCurrentUserInfo) {
+      await auxLoginFromUserInfo(localStorageCurrentUserInfo)
+      window.sessionStorage.setItem(CURRENT_USER_INFO_STR, localStorageCurrentUserInfo)
+    }
   }
 
-  isLoggedIn () {
-    // is logged in if the userAppDbId is known
-    // (either after register or after construction and load session info)
-    return !!this.userId
+  async _openDb () {
+    try {
+      this.userAppDb = common.utils.createPromisifiedHyperDB(this.userAppDbId)
```

Comment:
while we are at it, can you extract the used function from `common.utils` and `common.cypto` at the top of the file ?

Reply:
I prefer to leave it as it is more explicit and easier to read as the origin of the function is clear.